Device attach fails on volume creation due to duplicate bios.uuid

Copy-n-paste from email : 
```
    Hi,
    
    We are having some trouble getting the docker-volume-vsphere plugin to work. Do you see anything that we may be missing?
    
    The details and logs are below.
    
    Thanks!
    XXX

    Environment Details: 1 VM on RHEL 7.3, ESXi 6.0U1, Docker 17.05.0-ce, vDVS 0.14
    
    Steps to Reproduce:
    docker plugin install --grant-all-permissions --alias vsphere vmware/docker-volume-vsphere:0.14
    docker volume create --driver=vsphere --name=test
    
    Actual Result:
    Error response from daemon: create test: VolumeDriver.Create: Failed to create filesystem on /dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:1:0: exit status 1. Output = mke2fs 1.43.3 (04-Sep-2016)
    The file /dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:1:0 does not exist and no size was specified.
    
    Client Logs:
    2017-06-07 19:32:35.514944609 +0000 UTC [INFO] Starting plugin driver=vsphere log_level=info config="/etc/docker-volume-vsphere.conf"
    2017-06-07 19:32:35.515383164 +0000 UTC [INFO] Plugin options - port=1019
    2017-06-07 19:32:35.515411358 +0000 UTC [INFO] Getting volume data from unix:///var/run/docker.sock
    2017-06-07 19:32:35.539957187 +0000 UTC [INFO] Found 6 running or paused containers
    2017-06-07 19:32:35.546933529 +0000 UTC [INFO] Discovered 0 volumes in use.
    2017-06-07 19:32:35.546966645 +0000 UTC [INFO] Refcounting successfully completed
    2017-06-07 19:32:35.546986052 +0000 UTC [INFO] Docker VMDK plugin started version="vSphere Volume Driver v0.4" port=1019 mock_esx=false
    2017-06-07 19:32:35.54727681 +0000 UTC [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vsphere.sock"
    2017-06-07 19:07:25.347595558 +0000 UTC [DEBUG] vmdkOps.Get name=test
    2017-06-07 19:07:25.34765686 +0000 UTC [DEBUG] Run get request: version=
    2017-06-07 19:07:25.563177483 +0000 UTC [DEBUG] vmdkOp.Create name=test
    2017-06-07 19:07:25.563228464 +0000 UTC [DEBUG] Run get request: version=
    2017-06-07 19:07:25.816647424 +0000 UTC [INFO] Attaching volume and creating filesystem name=test fstype=ext4
    2017-06-07 19:07:25.818644055 +0000 UTC [DEBUG] vmdkOps.Attach name=test
    2017-06-07 19:07:25.818684762 +0000 UTC [DEBUG] Run get request: version=
    2017-06-07 19:07:44.342576236 +0000 UTC [WARNING] Exceeded timeout while waiting for device attach to completetimeout=10s device="/dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:1:0"
    2017-06-07 19:07:44.344666589 +0000 UTC [ERROR] Create filesystem failed, removing the volume name=test error="Failed to create filesystem on /dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:1:0: exit status 1. Output = mke2fs 1.43.3 (04-Sep-2016)\nThe file /dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:1:0 does not exist and no size was specified.\n"
    2017-06-07 19:07:44.344716477 +0000 UTC [DEBUG] vmdkOps.Detach name=test
    2017-06-07 19:07:44.344730957 +0000 UTC [DEBUG] Run get request: version=
    2017-06-07 19:07:47.108149631 +0000 UTC [DEBUG] vmdkOps.Remove name=test
    2017-06-07 19:07:47.10819686 +0000 UTC [DEBUG] Run get request: version=
    
    ESX Host Logs:
    06/07/17 23:07:39 1420822 [Thread-11] [INFO   ] cmd get with opts {} on tenant_uuid 11111111-1111-1111-1111-111111111111 datastore VNX01-T2-NFS-IT2 is allowed to execute
    06/07/17 23:07:39 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] executeRequest 'get' completed with ret={u'Error': 'Volume test not found (file: /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk)'}
    06/07/17 23:07:39 1420822 [Thread-12] [WARNING] Volume size not specified
    06/07/17 23:07:39 1420822 [Thread-12] [WARNING] Volume size not specified
    06/07/17 23:07:39 1420822 [Thread-12] [INFO   ] cmd create with opts {u'fstype': u'ext4'} on tenant_uuid 11111111-1111-1111-1111-111111111111 datastore VNX01-T2-NFS-IT2 is allowed to execute
    06/07/17 23:07:39 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] *** createVMDK: /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk opts = {u'fstype': u'ext4'}
    06/07/17 23:07:40 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [WARNING] Volume size not specified
    06/07/17 23:07:40 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] executeRequest 'create' completed with ret=None
    06/07/17 23:07:40 1420822 [Thread-13] [INFO   ] cmd attach with opts {} on tenant_uuid 11111111-1111-1111-1111-111111111111 datastore VNX01-T2-NFS-IT2 is allowed to execute
    06/07/17 23:07:40 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] *** attachVMDK: /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk to env-a-4vm1 (d66143b7-8ce0-4771-8bbe-a0feb6f8028d) VM uuid = 420a8c6b-e833-dbf8-f04c-390c5332d91f
    06/07/17 23:07:40 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] Attaching /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk as independent_persistent
    06/07/17 23:07:45 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] Disk /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk successfully attached. controller pci_slot_number=160, disk_slot=1
    06/07/17 23:07:45 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] executeRequest 'attach' completed with ret={'ControllerPciSlotNumber': '160', 'Unit': '1'}
    06/07/17 23:07:46 1420822 [Thread-14] [INFO   ] cmd detach with opts {} on tenant_uuid 11111111-1111-1111-1111-111111111111 datastore VNX01-T2-NFS-IT2 is allowed to execute
    06/07/17 23:07:46 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] *** detachVMDK: /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk from env-a-4vm1 (d66143b7-8ce0-4771-8bbe-a0feb6f8028d) VM uuid = 420a8c6b-e833-dbf8-f04c-390c5332d91f
    06/07/17 23:07:48 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] Disk detached /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk
    06/07/17 23:07:48 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] executeRequest 'detach' completed with ret=None
    06/07/17 23:07:48 1420822 [Thread-15] [INFO   ] cmd remove with opts {} on tenant_uuid 11111111-1111-1111-1111-111111111111 datastore VNX01-T2-NFS-IT2 is allowed to execute
    06/07/17 23:07:48 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] *** removeVMDK: /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk
    06/07/17 23:07:48 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] *** cleanVMDK: /vmfs/volumes/VNX01-T2-NFS-IT2/dockvols/11111111-1111-1111-1111-111111111111/test.vmdk
    06/07/17 23:07:48 1420822 [env-dsvm1 (45d97822-3f84-43c5-8585-0d0356ac0ed6)-[VNX01-T2-NFS-IT2] dockvols/11111111-1111-1111-1111-111111111111/test.vmdk] [INFO   ] executeRequest 'remove' completed with ret=None
    06/07/17 23:08:20 1421847 [MainThread] [INFO   ] Connecting to the local Service Instance as 'dcui'
    
```

Tried this out multiple times and the attach-wait times are anywhere between 6s - 10s. From debugging, the plugin starts its wait after the docker host kernel has completed attaching the device. There are  no obvious delays seen in the kernel logs:

Docker host kernel logs:
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.069868] vmw_pvscsi: msg type: 0x0 - MSG RING: 64/63 (5)
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.069873] vmw_pvscsi: msg: device added at scsi0:1:0
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.070082] scsi 2:0:1:0: Direct-Access     VMware   Virtual disk     1.0  PQ: 0 ANSI: 2
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.074260] sd 2:0:1:0: Attached scsi generic sg4 type 0
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.075813] sd 2:0:1:0: [sdd] 204800 512-byte logical blocks: (104 MB/100 MiB)
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.075840] sd 2:0:1:0: [sdd] Write Protect is off
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.075844] sd 2:0:1:0: [sdd] Mode Sense: 61 00 00 00
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.075869] sd 2:0:1:0: [sdd] Cache data unavailable
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.075871] sd 2:0:1:0: [sdd] Assuming drive cache: write through
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.076083] sd 2:0:1:0: [sdd] Cache data unavailable
Jun  8 21:01:42 hte-1s-eng-dhcp98 kernel: [60580.076086] sd 2:0:1:0: [sdd] Assuming drive cache: write through
Jun  8 21:01:43 hte-1s-eng-dhcp98 kernel: [60580.710503]  sdd: unknown partition table
Jun  8 21:01:43 hte-1s-eng-dhcp98 kernel: [60580.710716] sd 2:0:1:0: [sdd] Cache data unavailable
Jun  8 21:01:43 hte-1s-eng-dhcp98 kernel: [60580.710720] sd 2:0:1:0: [sdd] Assuming drive cache: write through
Jun  8 21:01:43 hte-1s-eng-dhcp98 kernel: [60580.710743] sd 2:0:1:0: [sdd] Attached SCSI disk

Plugin logs
2017-06-08 21:00:52.856006649 -0700 PDT [DEBUG] vmdkOp.Create name=y_5
2017-06-08 21:00:52.856329133 -0700 PDT [DEBUG] Run get request: version=
2017-06-08 21:01:24.248953919 -0700 PDT [INFO] Attaching volume and creating filesystem name="y_5" fstype=ext4
2017-06-08 21:01:24.249156484 -0700 PDT [DEBUG] vmdkOps.Attach name=y_5
2017-06-08 21:01:24.249226682 -0700 PDT [DEBUG] Run get request: version=
2017-06-08 21:01:56.483686652 -0700 PDT [ERROR] Starting wait for attachname="y_5" error=<nil>
2017-06-08 21:02:02.581062994 -0700 PDT [DEBUG] event: "/dev/disk/by-path/pci-0000:13:00.0-scsi-0:0:1:0": 0x100 == IN_CREATE
2017-06-08 21:02:02.581194605 -0700 PDT [INFO] Scan complete device="/dev/disk/by-path/pci-0000:13:00.0-scsi-0:0:1:0" event="/dev/disk/by-path/pci-0000:13:00.0-scsi-0:0:1:0": 0x100 == IN_CREATE

As seen above its taking around 6s to wait for the device to be detected by the plugin.

The issue is reproduced at times but from all the logs available doesn't seem to be an issue on the ESX side.

--- a/esx_service/vmdk_ops.py
+++ b/esx_service/vmdk_ops.py
@@ -164,8 +164,8 @@ def RunCommand(cmd):
 # for now we care about size and (maybe) policy
 def createVMDK(vmdk_path, vm_name, vol_name,
                opts={}, vm_uuid=None, tenant_uuid=None, datastore_url=None):
-    logging.info("*** createVMDK: %s opts = %s vm_name=%s vm_uuid=%s tenant_uuid=%s datastore_url=%s",
-                  vmdk_path, opts, vm_name, vm_uuid, tenant_uuid, datastore_url)
+    logging.info("*** createVMDK: %s opts=%s vm_name=%s vm_uuid=%s tenant_uuid=%s datastore_url=%s",
+                 vmdk_path, opts, vm_name, vm_uuid, tenant_uuid, datastore_url)
 
     if os.path.isfile(vmdk_path):
         # We are mostly here due to race or Plugin VMCI retry #1076
@@ -244,7 +244,8 @@ def createVMDK(vmdk_path, vm_name, vol_name,
 
 
 def cloneVMDK(vm_name, vmdk_path, opts={}, vm_uuid=None, datastore_url=None):
-    logging.info("*** cloneVMDK: %s opts = %s vm_uuid=%s datastore_url=%s", vmdk_path, opts, vm_uuid, datastore_url)
+    logging.info("*** cloneVMDK: %s opts = %s vm_uuid=%s datastore_url=%s",
+                 vmdk_path, opts, vm_uuid, datastore_url)
 
     # Get source volume path for cloning
     error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
@@ -656,7 +657,8 @@ def listVMDK(tenant):
             for x in vmdks]
 
 
-# Return VM managed object, reconnect if needed. Throws if fails twice.
+# Return VM managed object, reconnect if needed. Throws if connection fails twice.
+# returns None if the uuid is not found
 def findVmByUuid(vm_uuid):
     si = get_si()
     vm = si.content.searchIndex.FindByUuid(None, vm_uuid, True, False)
@@ -668,20 +670,34 @@ def vm_uuid2name(vm_uuid):
         return None
     return vm.config.name
 
-# Return error, or None for OK.
-def attachVMDK(vmdk_path, vm_uuid):
-    vm = findVmByUuid(vm_uuid)
-    logging.info("*** attachVMDK: %s to %s VM uuid = %s",
-                 vmdk_path, vm.config.name, vm_uuid)
-    return disk_attach(vmdk_path, vm)
+
+def attach_or_detach_VMDK(cmd, vmdk_path, vm_name, bios_uuid, vc_uuid):
+    """Attaches (if cmd="attach" or detaches a  VMDK.
+    Returns json reply to pass upstairs"""
+
+    logging.info("*** %s VMDK %s to VM '%s' , bios uuid = %s, VC uuid=%s)",
+                 cmd, vmdk_path, vm_name, vm_uuid, vc_uuid)
+    vm = findVmByUuid(vc_uuid)
+    if not vm:
+        logging.warning("Failed to find VM by VC UUID %s, trying BIOS UUID %s", vc_uuid, vm_uuid)
+        vm = findVmByUuid(bios_uuid)
+    if not vm:
+        msg = "Failed to find VM object for %s (%s %s)" % (vm_name, bios_uuid, vc_uuid)
+        logging.error(msg)
+        return err(msg)
+
+    if vm.config.name != vm_name:
+        log.warning("vm_name from vSocket '%s' does not match VM object '%s' ", vm_name, vm.config.name)
+
+    action = disk_attach if cmd == "attach" else disk_detach
+    return action(vmdk_path, vm)
 
 
 # Return error, or None for OK.
-def detachVMDK(vmdk_path, vm_uuid):
+def detachVMDK(vmdk_path, vm_uuid, vc_uuid):
     vm = findVmByUuid(vm_uuid)
-    logging.info("*** detachVMDK: %s from %s VM uuid = %s",
+    logging.info("*** detachVMDK: %s from %s VM uuid = %s (VC uuid=%s",
                  vmdk_path, vm.config.name, vm_uuid)
-    return disk_detach(vmdk_path, vm)
 
 
 # Check existence (and creates if needed) the path for docker volume VMDKs
@@ -831,7 +847,7 @@ def authorize_check(vm_uuid, datastore_url, cmd, opts, use_default_ds, datastore
 
 
 # gets the requests, calculates path for volumes, and calls the relevant handler
-def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+def executeRequest(vm_uuid, vc_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     """
     Executes a <cmd> request issused from a VM.
     The request is about volume <full_volume_name> in format volume@datastore.
@@ -871,8 +887,8 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     # default_datastore could be a real datastore name or a hard coded  one "_VM_DS"
     default_datastore = get_datastore_name(default_datastore_url)
 
-    logging.debug("executeRequest: vm uuid=%s name=%s, tenant_name=%s, default_datastore=%s",
-                  vm_uuid, vm_name, tenant_name, default_datastore)
+    logging.debug("executeRequest: vm uuid=%s VC uuid=%s name=%s, tenant_name=%s, default_datastore=%s",
+                  vm_uuid, vc_uuid, vm_name, tenant_name, default_datastore)
 
     if cmd == "list":
         threadutils.set_thread_name("{0}-nolock-{1}".format(vm_name, cmd))
@@ -902,7 +918,8 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
                   "default_datastore_url=%s datastore_url=%s",
                   vm_uuid, vm_name, tenant_uuid, tenant_name, default_datastore_url, datastore_url)
 
-    error_info = authorize_check(vm_uuid, datastore_url, cmd, opts, use_default_ds, datastore, vm_datastore)
+    error_info = authorize_check(vm_uuid, datastore_url, cmd, opts, use_default_ds, datastore,
+                                 vm_datastore)
     if error_info:
         return err(error_info)
 
@@ -950,12 +967,13 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
                                   datastore_url=datastore_url)
 
         # For attach/detach reconfigure tasks, hold a per vm lock.
-        elif cmd == "attach":
+        elif cmd == "attach" or cmd == "detach":
+            # TBD: get VM MOREF here and pass it down. Check that MO REF name / etc.
+            # marches with what we think baed on VSI
+            # on mismatch bail out with loud cry as a rename should impact all places.
+            # TBD: try to get it first based on VC.UUID and if it fails fall back on BIOS.UUID
             with lockManager.get_lock(vm_uuid):
-                response = attachVMDK(vmdk_path, vm_uuid)
-        elif cmd == "detach":
-            with lockManager.get_lock(vm_uuid):
-                response = detachVMDK(vmdk_path, vm_uuid)
+                response = attach_or_detach_VMDK(cmd=cmd, vmdk_path=vmdk_path, vm_name=vm_name, bios_uuid=vm_uuid, vc_uuid=vc_uuid)
         else:
             return err("Unknown command:" + cmd)
 
@@ -973,7 +991,7 @@ def connectLocalSi():
 
             # Connect to local server as user "dcui" since this is the Admin that does not lose its
             # Admin permissions even when the host is in lockdown mode. User "dcui" does not have a
-            # password - it is used by the ESXi local application DCUI (Direct Console User Interface)
+            # password - it is used by local application DCUI (Direct Console User Interface)
             # Version must be set to access newer features, such as VSAN.
             _service_instance = pyVim.connect.Connect(
                 host='localhost',
@@ -1535,11 +1553,13 @@ def execRequestThread(client_socket, cartel, request):
         group_info = vsi.get("/vm/%s/vmmGroupInfo" % vmm_leader)
         vm_name = group_info["displayName"]
         cfg_path = group_info["cfgPath"]
-        uuid = group_info["uuid"]
+        uuid = group_info["uuid"]            # BIOS UUID, see http://www.virtu-al.net/2015/12/04/a-quick-reference-of-vsphere-ids/
+        vc_uuid = group_info["vcUuid"]       # VC UUID
         # pyVmomi expects uuid like this one: 564dac12-b1a0-f735-0df3-bceb00b30340
         # to get it from uuid in VSI vms/<id>/vmmGroup, we use the following format:
         UUID_FORMAT = "{0}{1}{2}{3}-{4}{5}-{6}{7}-{8}{9}-{10}{11}{12}{13}{14}{15}"
         vm_uuid = UUID_FORMAT.format(*uuid.replace("-",  " ").split())
+        vc_uuid = UUID_FORMAT.format(*vc_uuid.replace("-",  " ").split())
 
         try:
             req = json.loads(request.decode('utf-8'))
@@ -1551,7 +1571,7 @@ def execRequestThread(client_socket, cartel, request):
             # If req from client does not include version number, set the version to
             # SERVER_PROTOCOL_VERSION by default to make backward compatible
             client_protocol_version = int(req["version"]) if "version" in req else SERVER_PROTOCOL_VERSION
-            logging.debug("execRequestThread: version=%d", client_protocol_version)
+            logging.debug("execRequestThread: client protocol version=%d", client_protocol_version)
             if client_protocol_version != SERVER_PROTOCOL_VERSION:
                 if client_protocol_version < SERVER_PROTOCOL_VERSION:
                     reply_string = err("vSphere Docker Volume Service client version ({}) is older than server version ({}), "
@@ -1560,11 +1580,13 @@ def execRequestThread(client_socket, cartel, request):
                     reply_string = err("vSphere Docker Volume Service client version ({}) is newer than server version ({}), "
                                     "please update the server.".format(client_protocol_version, SERVER_PROTOCOL_VERSION))
                 send_vmci_reply(client_socket, reply_string)
-                logging.info("executeRequest '%s' failed: %s", req["cmd"], reply_string)
+                logging.warning("executeRequest '%s' failed: %s", req["cmd"], reply_string)
                 return
 
             opts = req["details"]["Opts"] if "Opts" in req["details"] else {}
-            reply_string = executeRequest(vm_uuid=vm_uuid,
+            reply_string = executeRequest(
+                                vm_uuid=vm_uuid,
+                                vc_uuid=vc_uuid,
                                 vm_name=vm_name,
                                 config_path=cfg_path,
                                 cmd=req["cmd"],

--- a/esx_service/vmdk_ops.py
+++ b/esx_service/vmdk_ops.py
@@ -671,14 +671,20 @@ def vm_uuid2name(vm_uuid):
     return vm.config.name
 
 
-def attach_or_detach_VMDK(cmd, vmdk_path, vm_name, bios_uuid, vc_uuid):
+def attachVMDK(vmdk_path, vm_name, bios_uuid, vc_uuid):
+    return apply_action_VMDK(disk_attach, vmdk_path, vm_name, bios_uuid, vc_uuid)
+
+def detachVMDK(vmdk_path, vm_name, bios_uuid, vc_uuid):
+    return apply_action_VMDK(disk_detach, vmdk_path, vm_name, bios_uuid, vc_uuid)
+
+def apply_action_VMDK(action, vmdk_path, vm_name, bios_uuid, vc_uuid):
     # note: vc_uuid is the last one to avoid reworkign tests which use positional args and
     # not aware of vc_uuid
-    """Attaches (if cmd="attach" or detaches a  VMDK.
-    Returns json reply to pass upstairs"""
+    """Finds the VM and applies action(path,vm_MO) to it.
+    Returns json reply from action to pass upstairs, or json with 'err'"""
 
-    logging.info("*** %s VMDK %s to VM '%s' , bios uuid = %s, VC uuid=%s)",
-                 cmd, vmdk_path, vm_name, bios_uuid, vc_uuid)
+    logging.info("*** %s: VMDK %s to VM '%s' , bios uuid = %s, VC uuid=%s)",
+                 action.__name__, vmdk_path, vm_name, bios_uuid, vc_uuid)
     vm = None
     if vc_uuid:
         vm = findVmByUuid(vc_uuid)
@@ -693,7 +699,6 @@ def attach_or_detach_VMDK(cmd, vmdk_path, vm_name, bios_uuid, vc_uuid):
     if vm.config.name != vm_name:
         logging.warning("vm_name from vSocket '%s' does not match VM object '%s' ", vm_name, vm.config.name)
 
-    action = disk_attach if cmd == "attach" else disk_detach
     return action(vmdk_path, vm)
 
 
@@ -964,10 +969,14 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts, vc_u
                                   datastore_url=datastore_url)
 
         # For attach/detach reconfigure tasks, hold a per vm lock.
-        elif cmd == "attach" or cmd == "detach":
+        elif cmd == "attach":
             with lockManager.get_lock(vm_uuid):
-                response = attach_or_detach_VMDK(cmd=cmd, vmdk_path=vmdk_path, vm_name=vm_name,
-                                                 bios_uuid=vm_uuid, vc_uuid=vc_uuid)
+                response = attachVMDK(vmdk_path=vmdk_path, vm_name=vm_name,
+                                      bios_uuid=vm_uuid, vc_uuid=vc_uuid)
+        elif cmd == "detach":
+            with lockManager.get_lock(vm_uuid):
+                response = detachVMDK(vmdk_path=vmdk_path, vm_name=vm_name,
+                                      bios_uuid=vm_uuid, vc_uuid=vc_uuid)
         else:
             return err("Unknown command:" + cmd)
 

--- a/esx_service/vmdk_ops.py
+++ b/esx_service/vmdk_ops.py
@@ -1121,7 +1121,7 @@ def setStatusAttached(vmdk_path, vm, vm_dev_info=None):
     if not vol_meta:
         vol_meta = {}
     vol_meta[kv.STATUS] = kv.ATTACHED
-    vol_meta[kv.ATTACHED_VM_UUID] = vm.config.uuid
+    vol_meta[kv.ATTACHED_VM_UUID] = vm.config.instanceUuid
     vol_meta[kv.ATTACHED_VM_NAME] = vm.config.name
     if vm_dev_info:
         vol_meta[kv.ATTACHED_VM_DEV] = vm_dev_info
@@ -1176,19 +1176,19 @@ def getStatusAttached(vmdk_path):
     return attached, uuid, attach_as, vm_name
 
 def log_attached_volume(vmdk_path, kv_uuid, vol_name):
-       '''
-       Log appropriate message for volume thats already attached.
-       '''
-       cur_vm = findVmByUuid(kv_uuid)
+    '''
+    Log appropriate message for volume thats already attached.
+    '''
+    cur_vm = findVmByUuid(kv_uuid)
 
-       if cur_vm:
-          msg = "Disk {0} is already attached to VM {1}".format(vmdk_path,
-                                                             cur_vm.config.name)
-       else:
-          msg = "Failed to find VM {0}({1}), disk {2} is already attached".format(vol_name,
-                                                                                 kv_uuid,
-                                                                                 vmdk_path)
-       logging.warning(msg)
+    if cur_vm:
+        msg = "Disk {0} is already attached to VM {1}".format(vmdk_path,
+                                                              cur_vm.config.name)
+    else:
+        msg = "Failed to find VM {0}({1}), disk {2} is already attached".format(vol_name,
+                                                                                kv_uuid,
+                                                                                vmdk_path)
+    logging.warning(msg)
 
 def add_pvscsi_controller(vm, controllers, max_scsi_controllers, offset_from_bus_number):
     '''

--- a/esx_service/vmdk_ops.py
+++ b/esx_service/vmdk_ops.py
@@ -672,32 +672,29 @@ def vm_uuid2name(vm_uuid):
 
 
 def attach_or_detach_VMDK(cmd, vmdk_path, vm_name, bios_uuid, vc_uuid):
+    # note: vc_uuid is the last one to avoid reworkign tests which use positional args and
+    # not aware of vc_uuid
     """Attaches (if cmd="attach" or detaches a  VMDK.
     Returns json reply to pass upstairs"""
 
     logging.info("*** %s VMDK %s to VM '%s' , bios uuid = %s, VC uuid=%s)",
-                 cmd, vmdk_path, vm_name, vm_uuid, vc_uuid)
-    vm = findVmByUuid(vc_uuid)
-    if not vm:
-        logging.warning("Failed to find VM by VC UUID %s, trying BIOS UUID %s", vc_uuid, vm_uuid)
+                 cmd, vmdk_path, vm_name, bios_uuid, vc_uuid)
+    vm = None
+    if vc_uuid:
+        vm = findVmByUuid(vc_uuid)
+    if not vm: # either vc_uuid is not even passed, or we failed to find the VM by VC uuid:
+        logging.warning("Failed to find VM by VC UUID %s, trying BIOS UUID %s", vc_uuid, bios_uuid)
         vm = findVmByUuid(bios_uuid)
-    if not vm:
-        msg = "Failed to find VM object for %s (%s %s)" % (vm_name, bios_uuid, vc_uuid)
+    if not vm: # can find VM by VC or BIOS uuid
+        msg = "Failed to find VM object for %s (bios %s vc %s)" % (vm_name, bios_uuid, vc_uuid)
         logging.error(msg)
         return err(msg)
 
     if vm.config.name != vm_name:
-        log.warning("vm_name from vSocket '%s' does not match VM object '%s' ", vm_name, vm.config.name)
+        logging.warning("vm_name from vSocket '%s' does not match VM object '%s' ", vm_name, vm.config.name)
 
     action = disk_attach if cmd == "attach" else disk_detach
     return action(vmdk_path, vm)
-
-
-# Return error, or None for OK.
-def detachVMDK(vmdk_path, vm_uuid, vc_uuid):
-    vm = findVmByUuid(vm_uuid)
-    logging.info("*** detachVMDK: %s from %s VM uuid = %s (VC uuid=%s",
-                 vmdk_path, vm.config.name, vm_uuid)
 
 
 # Check existence (and creates if needed) the path for docker volume VMDKs
@@ -847,7 +844,7 @@ def authorize_check(vm_uuid, datastore_url, cmd, opts, use_default_ds, datastore
 
 
 # gets the requests, calculates path for volumes, and calls the relevant handler
-def executeRequest(vm_uuid, vc_uuid, vm_name, config_path, cmd, full_vol_name, opts):
+def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts, vc_uuid=None):
     """
     Executes a <cmd> request issused from a VM.
     The request is about volume <full_volume_name> in format volume@datastore.
@@ -968,12 +965,9 @@ def executeRequest(vm_uuid, vc_uuid, vm_name, config_path, cmd, full_vol_name, o
 
         # For attach/detach reconfigure tasks, hold a per vm lock.
         elif cmd == "attach" or cmd == "detach":
-            # TBD: get VM MOREF here and pass it down. Check that MO REF name / etc.
-            # marches with what we think baed on VSI
-            # on mismatch bail out with loud cry as a rename should impact all places.
-            # TBD: try to get it first based on VC.UUID and if it fails fall back on BIOS.UUID
             with lockManager.get_lock(vm_uuid):
-                response = attach_or_detach_VMDK(cmd=cmd, vmdk_path=vmdk_path, vm_name=vm_name, bios_uuid=vm_uuid, vc_uuid=vc_uuid)
+                response = attach_or_detach_VMDK(cmd=cmd, vmdk_path=vmdk_path, vm_name=vm_name,
+                                                 bios_uuid=vm_uuid, vc_uuid=vc_uuid)
         else:
             return err("Unknown command:" + cmd)
 

--- a/esx_service/vmdk_ops.py
+++ b/esx_service/vmdk_ops.py
@@ -1560,6 +1560,8 @@ def execRequestThread(client_socket, cartel, request):
                     reply_string = err("vSphere Docker Volume Service client version ({}) is newer than server version ({}), "
                                     "please update the server.".format(client_protocol_version, SERVER_PROTOCOL_VERSION))
                 send_vmci_reply(client_socket, reply_string)
+                logging.info("executeRequest '%s' failed: %s", req["cmd"], reply_string)
+                return
 
             opts = req["details"]["Opts"] if "Opts" in req["details"] else {}
             reply_string = executeRequest(vm_uuid=vm_uuid,

